### PR TITLE
Add a File field, like for .pdf uploads

### PR DESCRIPTION
**Pull request by kienstra**
# Changes
* Add a File field that allows non-image files:

ioabsdifadsf

* There's already an Image field, but that doesn't allow some files like `.pdf` files

Fixes [GF-3071](https://wpengine.atlassian.net/browse/GF-3071)

# Testing instructions
1. `cd /path/to/wp-content/plugins/genesis-custom-blocks/`

2. `git fetch`

3. `git checkout add/file-field`

4. `composer install && npm i && npm run build`

5. `/wp-admin` > Custom Blocks > Add New

6. Add a field with the type 'File': 
file-here

7. Click Publish

8. `/wp-admin` > Posts > Add New

9. Enter the name of the block you just created: 
thebisdf

10. In the block, click 'Upload' and upload any `.pdf` file

11. Expected: The file name displays: 
ioabsdifadsf

12. This won't preview the actual file in the Edit component, as the file could be a `.pdf`, a `.zip`, or any number of files

### CODE REVIEW

File: js/src/block-editor/controls/file.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/**
```
Comment (kienstra) [inline js/src/block-editor/controls/file.js:1]: This file is mainly copied from `image.js`, so they could probably use a common component that they both call.

Comment (kienstra) [reply js/src/block-editor/controls/file.js:1]: Actually, there are enough differences that a common component might make it more complex.

File: package-lock.json [hunk 2]
```diff
@@ -5134,6 +5134,12 @@
           "integrity": "sha1-zg6+ql94yxiSXqfYENe1mwEP1RU=",
           "dev": true
         },
+        "prettier": {
+          "version": "npm:wp-prettier@2.2.1-beta-1",
+          "resolved": "https://registry.npmjs.org/wp-prettier/-/wp-prettier-2.2.1-beta-1.tgz",
+          "integrity": "sha512-+JHkqs9LC/JPp51yy1hzs3lQ7qeuWCwOcSzpQNeeY/G7oSpnF61vxt7hRh87zNRTr6ob2ndy0W8rVzhgrcA+Gw==",
+          "dev": true
+        },
```
Comment (kienstra) [inline package-lock.json]: Not related to this PR, but it's annoying to see this in the diff every time you do `npm install`

File: php/Blocks/Loader.php [hunk 3]
```diff
@@ -72,7 +72,7 @@ public function init() {
 	 */
 	public function register_hooks() {
 		add_action( 'enqueue_block_editor_assets', [ $this, 'editor_assets' ] );
-		add_filter( 'block_categories', [ $this, 'register_categories' ] );
+		add_filter( 'block_categories_all', [ $this, 'register_categories' ] );
```
Comment (kienstra) [inline php/Blocks/Loader.php]: This will need to go out before WP `5.8` to prevent the PHP notice.

File: js/src/block-editor/hooks/useMedia.js [hunk 4]
```diff
@@ -43,67 +43,64 @@ import { __, sprintf } from '@wordpress/i18n';
...
+		} else if ( 'string' === typeof media ) {
 			// Backwards-compatibility: the fieldValue used to be the URL, not the ID.
-			setImageSrc( newImage );
+			setMediaSrc( media );
 		}
 
-		if ( newImage?.alt ) {
-			setImageAlt( newImage.alt );
-		} else if ( newImage?.source_url ) { // eslint-disable-line camelcase
+		if ( media?.alt ) {
```
Comment (kienstra) [inline js/src/block-editor/hooks/useMedia.js:69]: Most of this file's diff is just renaming parameters. 

Like here, it changes `newImage` to `media`, as this hook is now used for files and images.

**Review by dreamwhisper — APPROVED** [review]

**Comment by johnstonphilip** [comment]
Nice work Ryan! This is working well for me when logged in as an administrator. However, if I log in as a contributor, I see the file field, but it doesn't actually upload the file. This isn't a `must-fix` right now, but possibly something to add in the future, if the user doesn't actually have permission to upload files. Perhaps we shouldn't show the upload field to them. 

Screen Shot 2021-07-19 at 3 07 16 PM


The main thing I wanted to be sure of was user capabilities work, which they do. 

I was reading that you need to wrap `MediaUpload` in `MediaUploadCheck` here:
https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/media-upload/README.md

**Review by johnstonphilip — APPROVED** [review]

**Comment by kienstra** [comment]
Thanks, @dreamwhisper! 

Great idea to test it with a contributor, never though of that. Also, good idea to use a `<MediaUploadCheck>`

**Comment by kienstra** [comment]
If it's OK, I [bumped the WP Tested Up To version to 5.8](https://github.com/studiopress/genesis-custom-blocks/pull/74/commits/e798b9e4be98a42f6e9cbc46d726baf3ee6bde8c), forgot to do that in #85

**Pull request merged**